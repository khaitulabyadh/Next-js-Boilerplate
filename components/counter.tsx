'use client'

import { useState } from 'react'

import { Button } from '@/components/ui/button'

export const Counter = () => {
  const [count, setCount] = useState(0)

  return (
    <div className="mb-4">
      <span className="mr-4">{count}</span>
      <Button onClick={() => setCount((x) => x + 1)}>{count} + 1</Button>
    </div>
  )
}
