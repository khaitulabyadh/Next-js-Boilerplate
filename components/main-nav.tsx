'use client'
// todo => use server?
import Image from 'next/image'
import Link from 'next/link'
import { usePathname } from 'next/navigation'

import { cn } from '@/lib/utils'

export function MainNav() {
  const pathname = usePathname()

  return (
    <div className="mr-4 md:flex">
      <Link
        href="/"
        className="mr-6 flex items-center"
      >
        <Image
          src="/next.svg"
          alt="Next.js Logo"
          className="inline-block h-6 w-16 dark:invert"
          width={100}
          height={24}
        />
        <span className="ml-1 inline-block font-bold">14 Boilerplate</span>
      </Link>
      <nav className="hidden items-center gap-4 text-sm sm:flex lg:gap-6">
        <Link
          href="/loading-and-streaming"
          className={cn(
            'transition-colors hover:text-foreground/80',
            pathname === '/loading-and-streaming'
              ? 'text-foreground'
              : 'text-foreground/60',
          )}
        >
          Loading and streaming
        </Link>
        <Link
          href="/dashboard"
          className={cn(
            'transition-colors hover:text-foreground/80',
            pathname?.startsWith('/dashboard')
              ? 'text-foreground'
              : 'text-foreground/60',
          )}
        >
          dashboard
        </Link>
        <Link
          href="https://github.com/qinsong77"
          className={cn(
            'hidden text-foreground/60 transition-colors hover:text-foreground/80 lg:block',
          )}
        >
          GitHub
        </Link>
      </nav>
    </div>
  )
}
