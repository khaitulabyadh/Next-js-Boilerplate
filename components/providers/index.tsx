'use client'

import { ThemeProvider } from '@/components/providers/theme-provider'
import { Toaster } from '@/components/ui/toaster'
import { TooltipProvider } from '@/components/ui/tooltip'

export function Providers({ children }: { children: React.ReactNode }) {
  return (
    <>
      <ThemeProvider
        attribute="class"
        defaultTheme="system"
        enableSystem
        disableTransitionOnChange
      >
        <Toaster />
        <TooltipProvider>{children}</TooltipProvider>
      </ThemeProvider>
    </>
  )
}
