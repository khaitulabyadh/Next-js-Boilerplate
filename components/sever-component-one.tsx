'use server'

import { readFile } from 'node:fs/promises'

import { logger } from '@/lib/shared'

export const SeverComponentOne = async (properties: { value: string }) => {
  logger.info(properties, 'SeverComponentOne props')
  const tsconfig = await readFile('tsconfig.json', 'utf8')

  // await new Promise((resolve) => setTimeout(resolve, 500));

  return (
    <div>
      My server component2 with val: {properties.value}
      <h4 className="scroll-m-20 text-xl font-semibold tracking-tight">
        show tsconfig.json from server:
      </h4>
      <div className="mt-4 max-w-3xl">
        <code>{tsconfig}</code>
      </div>
    </div>
  )
}
