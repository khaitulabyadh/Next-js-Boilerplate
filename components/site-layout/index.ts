export * from './external-link'
export * from './site-footer'
export * from './site-header'
