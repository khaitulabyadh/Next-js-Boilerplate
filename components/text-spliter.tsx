'use client'
import { startTransition, useState } from 'react'

import { useCustomerInterval } from '@/hooks/client'

const orgText = 'React & Next.js 14'
export const TextSplitter = () => {
  const [count, setCount] = useState(0)
  const pausedTimeRef = useCustomerInterval(() => {
    startTransition(() => {
      setCount((prevState) => {
        if (prevState === orgText.length) {
          return 0
        }
        const val = prevState + 1
        if (val === orgText.length - 1) {
          pausedTimeRef.current = 1000
        }
        return val
      })
    })
  }, 100)
  return (
    <div>
      <h1 className="mb-4 text-2xl font-bold leading-normal md:text-4xl">
        This is a Next.js Boilerplate
        <br />
        Base on <span>{orgText.slice(0, count)}</span>
        {count > 0 && (
          <span
            style={{ verticalAlign: '0.25rem' }}
            className="ml-1 text-2xl"
            aria-hidden="true"
          >
            |
          </span>
        )}
      </h1>
    </div>
  )
}
