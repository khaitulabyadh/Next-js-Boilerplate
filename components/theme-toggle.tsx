'use client'

import { Apple, Moon, MoonIcon, Sun, SunIcon } from 'lucide-react'
import { useTheme } from 'next-themes'
import * as React from 'react'

import { Button } from '@/components/ui/button'
import {
  DropdownMenu,
  DropdownMenuContent,
  DropdownMenuItem,
  DropdownMenuTrigger,
} from '@/components/ui/dropdown-menu'
import { logger } from '@/lib/client'

export function ThemeToggle() {
  const { setTheme } = useTheme()

  return (
    <DropdownMenu>
      <DropdownMenuTrigger asChild>
        <Button
          variant="ghost"
          className="w-9 px-0"
        >
          <SunIcon className="size-[1.2rem] rotate-0 scale-100 transition-all dark:-rotate-90 dark:scale-0" />
          <MoonIcon className="absolute size-[1.2rem] rotate-90 scale-0 transition-all dark:rotate-0 dark:scale-100" />
          <span className="sr-only">Toggle theme</span>
        </Button>
      </DropdownMenuTrigger>
      <DropdownMenuContent align="end">
        <DropdownMenuItem
          onClick={() => {
            setTheme('light')
            logger.info('theme changed to light')
          }}
        >
          <Sun className="mr-2 size-4 transition-all " />
          Light
        </DropdownMenuItem>
        <DropdownMenuItem onClick={() => setTheme('dark')}>
          <Moon className="mr-2 size-4 transition-all" />
          Dark
        </DropdownMenuItem>
        <DropdownMenuItem onClick={() => setTheme('system')}>
          <Apple className="mr-2 size-4 transition-all" />
          System
        </DropdownMenuItem>
      </DropdownMenuContent>
    </DropdownMenu>
  )
}
