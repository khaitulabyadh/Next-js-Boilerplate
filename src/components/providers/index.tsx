'use client'

import { ThemeProvider } from '@/components/providers/theme-provider'
import { TooltipProvider } from '@/components/ui/tooltip'

export function Providers({ children }: { children: React.ReactNode }) {
  return (
    <>
      <ThemeProvider
        attribute="class"
        defaultTheme="system"
        enableSystem
        disableTransitionOnChange
      >
        <TooltipProvider>{children}</TooltipProvider>
      </ThemeProvider>
    </>
  )
}
