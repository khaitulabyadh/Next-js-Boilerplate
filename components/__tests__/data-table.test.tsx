import { type ColumnDef } from '@tanstack/react-table'
import { render } from '@testing-library/react'
import { describe, expect, it } from 'vitest'

import { DataTable, DataTableColumnHeader } from '@/components/data-table'

type Person = {
  id: string
  name: string
  email: string
  role: string
}

const columns: ColumnDef<Person>[] = [
  {
    accessorKey: 'name',
    header: ({ column }) => (
      <DataTableColumnHeader
        column={column}
        title="Name"
      />
    ),
  },
  {
    accessorKey: 'email',
    header: ({ column }) => (
      <DataTableColumnHeader
        column={column}
        title="Email"
      />
    ),
    enableSorting: false,
  },
  {
    accessorKey: 'role',
    header: ({ column }) => (
      <DataTableColumnHeader
        column={column}
        title="Role"
      />
    ),
  },
]

const data: Person[] = [
  { id: '1', name: 'John Doe', email: 'john@example.com', role: 'Admin' },
  { id: '2', name: 'Jane Smith', email: 'jane@example.com', role: 'User' },
  { id: '3', name: 'Bob Johnson', email: 'bob@example.com', role: 'Manager' },
  { id: '4', name: 'Alice Brown', email: 'alice@example.com', role: 'User' },
]

describe('[component]-DataTable', () => {
  it('renders correctly', () => {
    const { asFragment } = render(
      <DataTable
        columns={columns}
        data={data}
      />,
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
