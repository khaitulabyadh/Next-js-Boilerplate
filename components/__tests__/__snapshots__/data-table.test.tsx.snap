// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`[component]-DataTable > renders correctly 1`] = `
<DocumentFragment>
  <div
    class="grid"
  >
    <div
      class="relative overflow-hidden border-b"
      dir="ltr"
      style="position: relative; --radix-scroll-area-corner-width: 0px; --radix-scroll-area-corner-height: 0px;"
    >
      <style>
        [data-radix-scroll-area-viewport]{scrollbar-width:none;-ms-overflow-style:none;-webkit-overflow-scrolling:touch;}[data-radix-scroll-area-viewport]::-webkit-scrollbar{display:none}
      </style>
      <div
        class="h-full w-full rounded-[inherit]"
        data-radix-scroll-area-viewport=""
        style="overflow-x: scroll; overflow-y: scroll;"
      >
        <div
          style="min-width: 100%; display: table;"
        >
          <table
            class="caption-bottom text-sm w-full table-auto text-center"
          >
            <thead
              class="[&_tr]:border-b sticky top-0 z-10 bg-secondary"
            >
              <tr
                class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted border-none"
              >
                <th
                  class="h-12 px-4 align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 relative text-center before:absolute before:inset-y-1/4 before:right-0 before:w-px before:bg-gray-300 before:content-[""]"
                  colspan="1"
                >
                  <div
                    class="flex items-center justify-center space-x-1.5"
                  >
                    <div
                      class="w-full truncate"
                    >
                      Name
                    </div>
                    <div
                      class="flex flex-col"
                    >
                      <button
                        class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent hover:text-accent-foreground rounded-md h-3 w-4 p-0"
                        data-state="closed"
                      >
                        <svg
                          class="lucide lucide-chevron-up size-3"
                          fill="none"
                          height="24"
                          stroke="currentColor"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                          stroke-width="2"
                          viewBox="0 0 24 24"
                          width="24"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="m18 15-6-6-6 6"
                          />
                        </svg>
                        <span
                          class="sr-only"
                        >
                          升序排序
                        </span>
                      </button>
                      <button
                        class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent hover:text-accent-foreground rounded-md h-3 w-4 p-0"
                        data-state="closed"
                      >
                        <svg
                          class="lucide lucide-chevron-down size-3"
                          fill="none"
                          height="24"
                          stroke="currentColor"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                          stroke-width="2"
                          viewBox="0 0 24 24"
                          width="24"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="m6 9 6 6 6-6"
                          />
                        </svg>
                        <span
                          class="sr-only"
                        >
                          降序排序
                        </span>
                      </button>
                    </div>
                  </div>
                </th>
                <th
                  class="h-12 px-4 align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 relative text-center before:absolute before:inset-y-1/4 before:right-0 before:w-px before:bg-gray-300 before:content-[""]"
                  colspan="1"
                >
                  <div
                    class="w-full text-center"
                  >
                    Email
                  </div>
                </th>
                <th
                  class="h-12 px-4 align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 relative text-center"
                  colspan="1"
                >
                  <div
                    class="flex items-center justify-center space-x-1.5"
                  >
                    <div
                      class="w-full truncate"
                    >
                      Role
                    </div>
                    <div
                      class="flex flex-col"
                    >
                      <button
                        class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent hover:text-accent-foreground rounded-md h-3 w-4 p-0"
                        data-state="closed"
                      >
                        <svg
                          class="lucide lucide-chevron-up size-3"
                          fill="none"
                          height="24"
                          stroke="currentColor"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                          stroke-width="2"
                          viewBox="0 0 24 24"
                          width="24"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="m18 15-6-6-6 6"
                          />
                        </svg>
                        <span
                          class="sr-only"
                        >
                          升序排序
                        </span>
                      </button>
                      <button
                        class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent hover:text-accent-foreground rounded-md h-3 w-4 p-0"
                        data-state="closed"
                      >
                        <svg
                          class="lucide lucide-chevron-down size-3"
                          fill="none"
                          height="24"
                          stroke="currentColor"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                          stroke-width="2"
                          viewBox="0 0 24 24"
                          width="24"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="m6 9 6 6 6-6"
                          />
                        </svg>
                        <span
                          class="sr-only"
                        >
                          降序排序
                        </span>
                      </button>
                    </div>
                  </div>
                </th>
              </tr>
            </thead>
            <tbody
              class="[&_tr:last-child]:border-0"
            >
              <tr
                class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted h-12"
                data-state="false"
              >
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0 text-center"
                >
                  John Doe
                </td>
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0 text-center"
                >
                  john@example.com
                </td>
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0 text-center"
                >
                  Admin
                </td>
              </tr>
              <tr
                class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted h-12"
                data-state="false"
              >
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0 text-center"
                >
                  Jane Smith
                </td>
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0 text-center"
                >
                  jane@example.com
                </td>
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0 text-center"
                >
                  User
                </td>
              </tr>
              <tr
                class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted h-12"
                data-state="false"
              >
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0 text-center"
                >
                  Bob Johnson
                </td>
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0 text-center"
                >
                  bob@example.com
                </td>
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0 text-center"
                >
                  Manager
                </td>
              </tr>
              <tr
                class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted h-12"
                data-state="false"
              >
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0 text-center"
                >
                  Alice Brown
                </td>
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0 text-center"
                >
                  alice@example.com
                </td>
                <td
                  class="p-4 align-middle [&:has([role=checkbox])]:pr-0 text-center"
                >
                  User
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
